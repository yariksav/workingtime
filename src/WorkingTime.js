
export default class WorkingTime {
  constructor (periods) {
    this.periods = this.setPeriods(periods)
  }
  
  setPeriods (periods) {
    return (periods || []).map(item => {
      let indexTo = item.close ? item.close.day + item.close.time : '00000'
      return {
        indexFrom: item.open.day + item.open.time,
        indexTo,
        ...item
      }
    }).sort((a, b) => a.indexFrom > b.indexFrom)
  }

  isOpen () {
    return this.isOpenAt(new Date())
  }

  isOpenAt (date) {
    if (!date || !(date instanceof Date)) {
      return false
    }
    return Boolean(this.currentPeriodFor(date))
  }

  // isNonStop () {
  // }

  static getDateIndex (date) {
    if (!date) {
      return null
    }
    let currentDay = date.getDay()
    let currentTime = ('00' + date.getHours()).slice(-2) + ('00' + date.getMinutes()).slice(-2)
    return currentDay + currentTime
  }
  
  static fetchDate (date, event) {
    if (!event || !(date instanceof Date)) {
      return null
    }
    const day = date.getDay()
    const daysDiff = event.day >= day ? event.day - day : event.day + 7 - day
    const hours = event.time.slice(0, 2)
    const minutes = event.time.slice(2, 4)
    return new Date(date.getFullYear(), date.getMonth(), date.getDate() + daysDiff, hours, minutes, 0)
  }

  getInfo () {
    return this.getInfoFor(new Date())
  }

  static getDateDiffs (date1, date2) {
    if (!date1 || !date2) {
      return null
    }
    return (date2.getTime() - date1.getTime()) / 1000
  }

  getInfoFor (date) {
    const period = this.currentPeriodFor(date)
    const nextPeriod = this.nextPeriodFor(date)
    const isOpen = Boolean(period)
    const nextEventDate = isOpen ? WorkingTime.fetchDate(date, period.close) : WorkingTime.fetchDate(date, nextPeriod && nextPeriod.open)
    return {
      isOpen,
      currentPeriod: period,
      currentDate: date,
      nextPeriod,
      nextDiff: WorkingTime.getDateDiffs(date, nextEventDate),
      nextEventDate,
      nextEventType: isOpen ? 'close' : 'open'
    }
  }

  nextPeriodFor (date) {
    let currentIndex = WorkingTime.getDateIndex(date)
    let firstElement = null
    let period = this.periods.reduce((element, item) => {
      if (item.indexFrom > currentIndex && (!element || element.indexFrom >= item.indexFrom)) {
        element = item
      }
      return element
    }, null)
    if (!period) {
      period = this.periods[0]
    }
    return period
  }

  currentPeriodFor (date) {
    let currentIndex = WorkingTime.getDateIndex(date)
    let period = this.periods.find(item => {
      return item.indexFrom < item.indexTo ? 
        item.indexFrom <= currentIndex && item.indexTo >= currentIndex :
        item.indexFrom <= currentIndex || item.indexTo >= currentIndex
    })
    return period
  }
}
