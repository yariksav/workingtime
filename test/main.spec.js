import WorkingTime from '../src/WorkingTime'

const genDate = (day, hour, minute = 0) => {
  return new Date (0, 0, day, hour, minute)
}

describe('main', () => {
  it ('should gen correct date', () => {
    expect(genDate(0, 0).getDay()).toBe(0)
    expect(genDate(1, 0).getDay()).toBe(1)
    expect(genDate(1, 14).getHours()).toBe(14)
    expect(genDate(1, 13, 12).getMinutes()).toBe(12)
  })

  it('should set periods', () => {
    const testPeriod = []
    const wt = new WorkingTime(testPeriod)
    expect(wt.periods).toEqual(testPeriod)
  })

  it('should be correct openAt results for simple period', () => {
    const periods = [
      { open: { day: 0, time: '1000' }, close: { day: 0, time: '1300' } },
      { open: { day: 1, time: '0900' }, close: { day: 1, time: '2300' } },
      { open: { day: 2, time: '0900' }, close: { day: 2, time: '2300' } },
      { open: { day: 3, time: '0900' }, close: { day: 3, time: '2300' } },
      { open: { day: 4, time: '0900' }, close: { day: 4, time: '2300' } },
      { open: { day: 5, time: '0900' }, close: { day: 5, time: '2200' } },
      { open: { day: 6, time: '0900' }, close: { day: 6, time: '1700' } }
    ]
    const wt = new WorkingTime(periods)
    expect(wt.isOpenAt(genDate(0, 0))).toBeFalsy()
    expect(wt.isOpenAt(genDate(0, 10))).toBeTruthy()
    expect(wt.isOpenAt(genDate(0, 13))).toBeTruthy()
    expect(wt.isOpenAt(genDate(0, 14))).toBeFalsy()
    expect(wt.isOpenAt(genDate(0, 24))).toBeFalsy()
    expect(wt.isOpenAt(genDate(1, 8))).toBeFalsy()
    expect(wt.isOpenAt(genDate(1, 10))).toBeTruthy()
    expect(wt.isOpenAt(genDate(1, 24))).toBeFalsy()
    expect(wt.isOpenAt(genDate(2, 1))).toBeFalsy()
    expect(wt.isOpenAt(genDate(2, 23))).toBeTruthy()
    expect(wt.isOpenAt(genDate(6, 16))).toBeTruthy()
    expect(wt.isOpenAt(genDate(6, 18))).toBeFalsy()

    expect(wt.nextPeriodFor(genDate(0, 20))).toMatchObject({ open: { day: 1, time: '0900' }, close: { day: 1, time: '2300' } })
    expect(wt.nextPeriodFor(genDate(6, 20))).toMatchObject({ open: { day: 0, time: '1000' }, close: { day: 0, time: '1300' } })
  })


  it('should be correct openAt results for complex period', () => {
    const periods = [
      { open: { day: 1, time: '0900' }, close: { day: 2, time: '0200' } },
      { open: { day: 2, time: '0900' }, close: { day: 3, time: '0200' } },
      { open: { day: 3, time: '0900' }, close: { day: 4, time: '0200' } },
      { open: { day: 0, time: '0900' }, close: { day: 0, time: '2300' } }, // array should not be sorted by days
      { open: { day: 4, time: '0900' }, close: { day: 5, time: '0200' } },
      { open: { day: 5, time: '0900' }, close: { day: 5, time: '2300' } },
      { open: { day: 6, time: '0900' }, close: { day: 0, time: '0500' } }
    ]
    const wt = new WorkingTime(periods)
    expect(wt.isOpenAt(genDate(0, 0))).toBeTruthy()
    expect(wt.isOpenAt(genDate(0, 10))).toBeTruthy()
    expect(wt.isOpenAt(genDate(0, 23))).toBeTruthy()
    expect(wt.isOpenAt(genDate(0, 24))).toBeFalsy()
    expect(wt.isOpenAt(genDate(1, 8))).toBeFalsy()
    expect(wt.isOpenAt(genDate(1, 10))).toBeTruthy()
    expect(wt.isOpenAt(genDate(1, 24))).toBeTruthy()
    expect(wt.isOpenAt(genDate(2, 0))).toBeTruthy()
    expect(wt.isOpenAt(genDate(2, 1))).toBeTruthy()
    expect(wt.isOpenAt(genDate(2, 3))).toBeFalsy()
    expect(wt.isOpenAt(genDate(2, 9))).toBeTruthy()

    expect(wt.nextPeriodFor(genDate(6, 20))).toMatchObject({ open: { day: 0, time: '0900' }, close: { day: 0, time: '2300' } })
    expect(wt.nextPeriodFor(genDate(0, 0))).toMatchObject({ open: { day: 0, time: '0900' }, close: { day: 0, time: '2300' } })
    expect(wt.nextPeriodFor(genDate(0, 20))).toMatchObject({ open: { day: 1, time: '0900' }, close: { day: 2, time: '0200' } })

    expect(wt.getInfoFor(genDate(0, 20))).toMatchSnapshot()
    expect(wt.getInfoFor(genDate(1, 0))).toMatchSnapshot()
    expect(wt.getInfoFor(genDate(6, 23))).toMatchSnapshot()
  })

  it('should be correct openAt results for complex period', () => {
    const periods = [
      { open: { day: 3, time: '0900' }, close: { day: 4, time: '0200' } },
      { open: { day: 4, time: '0900' }, close: { day: 5, time: '0200' } },
      { open: { day: 5, time: '0900' }, close: { day: 2, time: '2300' } }
    ]
    const wt = new WorkingTime(periods)
    expect(wt.isOpenAt(genDate(2, 10))).toBeTruthy()
  })

  it('should be correct openAt results for non stop', () => {
    const periods = [
      { open: { day: 0, time: '0000' } }
    ]
    const wt = new WorkingTime(periods)
    expect(wt.isOpenAt(genDate(0, 0))).toBeTruthy()
    expect(wt.isOpenAt(genDate(1, 10))).toBeTruthy()
    expect(wt.isOpenAt(genDate(6, 24))).toBeTruthy()
    expect(wt.getInfoFor(genDate(2, 23))).toMatchSnapshot()
  })

  it('should be correct openAt results for none period', () => {
    const wt = new WorkingTime()
    expect(wt.isOpenAt(genDate(0, 0))).toBeFalsy()
    expect(wt.isOpenAt(genDate(1, 10))).toBeFalsy()
    expect(wt.isOpenAt(genDate(6, 24))).toBeFalsy()
    expect(wt.getInfoFor(genDate(2, 23))).toMatchSnapshot()
  })
})