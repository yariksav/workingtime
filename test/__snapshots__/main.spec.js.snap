// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main should be correct openAt results for complex period 1`] = `
Object {
  "currentDate": 1899-12-31T19:00:00.000Z,
  "currentPeriod": Object {
    "close": Object {
      "day": 0,
      "time": "2300",
    },
    "indexFrom": "00900",
    "indexTo": "02300",
    "open": Object {
      "day": 0,
      "time": "0900",
    },
  },
  "isOpen": true,
  "nextDiff": 10800,
  "nextEventDate": 1899-12-31T22:00:00.000Z,
  "nextEventType": "close",
  "nextPeriod": Object {
    "close": Object {
      "day": 2,
      "time": "0200",
    },
    "indexFrom": "10900",
    "indexTo": "20200",
    "open": Object {
      "day": 1,
      "time": "0900",
    },
  },
}
`;

exports[`main should be correct openAt results for complex period 2`] = `
Object {
  "currentDate": 1899-12-31T23:00:00.000Z,
  "currentPeriod": undefined,
  "isOpen": false,
  "nextDiff": 32400,
  "nextEventDate": 1900-01-01T08:00:00.000Z,
  "nextEventType": "open",
  "nextPeriod": Object {
    "close": Object {
      "day": 2,
      "time": "0200",
    },
    "indexFrom": "10900",
    "indexTo": "20200",
    "open": Object {
      "day": 1,
      "time": "0900",
    },
  },
}
`;

exports[`main should be correct openAt results for complex period 3`] = `
Object {
  "currentDate": 1900-01-06T22:00:00.000Z,
  "currentPeriod": Object {
    "close": Object {
      "day": 0,
      "time": "0500",
    },
    "indexFrom": "60900",
    "indexTo": "00500",
    "open": Object {
      "day": 6,
      "time": "0900",
    },
  },
  "isOpen": true,
  "nextDiff": 21600,
  "nextEventDate": 1900-01-07T04:00:00.000Z,
  "nextEventType": "close",
  "nextPeriod": Object {
    "close": Object {
      "day": 0,
      "time": "2300",
    },
    "indexFrom": "00900",
    "indexTo": "02300",
    "open": Object {
      "day": 0,
      "time": "0900",
    },
  },
}
`;

exports[`main should be correct openAt results for non stop 1`] = `
Object {
  "currentDate": 1900-01-02T22:00:00.000Z,
  "currentPeriod": Object {
    "indexFrom": "00000",
    "indexTo": "00000",
    "open": Object {
      "day": 0,
      "time": "0000",
    },
  },
  "isOpen": true,
  "nextDiff": null,
  "nextEventDate": null,
  "nextEventType": "close",
  "nextPeriod": Object {
    "indexFrom": "00000",
    "indexTo": "00000",
    "open": Object {
      "day": 0,
      "time": "0000",
    },
  },
}
`;

exports[`main should be correct openAt results for none period 1`] = `
Object {
  "currentDate": 1900-01-02T22:00:00.000Z,
  "currentPeriod": undefined,
  "isOpen": false,
  "nextDiff": null,
  "nextEventDate": null,
  "nextEventType": "open",
  "nextPeriod": undefined,
}
`;
